solr/core/src/java/org/apache/solr/search/MutableBitDocSet.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.search;
+
+import org.apache.lucene.util.FixedBitSet;
+
+public class MutableBitDocSet extends BitDocSet {
+  public MutableBitDocSet(FixedBitSet bits) {
+    super(bits);
+  }
+
+  @Override
+  public DocSet andNot(DocSet other) {
+    if (other instanceof BitDocSet) {
+      bits.andNot(((BitDocSet) other).bits);
+    } else {
+      DocIterator iter = other.iterator();
+      while (iter.hasNext()) {
+        int doc = iter.nextDoc();
+        if (doc < bits.length()) {
+          bits.clear(doc);
+        }
+      }
+    }
+    return new MutableBitDocSet(bits);
```

Comment:
Need comment - can't return `this` since `BitSetDoc` stores size cached. Easier to just create new lightweight object.

Reply:
addressed in bc97db7183cedabbb7615c081298f45f34750bb1